odoo/addons/base/models/ir_fields.py

```diff
@@ -383,6 +383,34 @@ def db_id_for(self, model, field, subfield, value):
                 {'moreinfo': action})
         return id, field_type, warnings
 
+    def _xmlid_to_record_id(self, xmlid, model):
+        """ Return the record id corresponding to the given external id,
+        provided that the record actually exists; otherwise return ``None``.
+        """
+        import_cache = self.env.context.get('import_cache', {})
+        result = import_cache.get(xmlid)
+        if result is not None:
+            if result[0] != model._name:
+                raise ValueError("External ID %s not with model %r" % (xmlid, model._name))
+            return result[1] or None
+
+        module, name = xmlid.split('.', 1)
+        query = """
+            SELECT d.model, d.res_id
+            FROM ir_model_data d
+            JOIN "{}" r ON d.res_id = r.id
+            WHERE d.module = %s AND d.name = %s
+        """.format(model._table)
+        self.env.cr.execute(query, [module, name])
+        result = self.env.cr.fetchone()
+
+        if result is not None:
+            import_cache[xmlid] = result
+            if result[0] != model._name:
+                raise ValueError("External ID %s not with model %r" % (xmlid, model._name))
```

Comment:
Maybe also show what's the actual model of the xid? e.g. `External ID %r invalid: expected model %r, found %r`.